### ISSUE TITLE
🦠 Model Request: Tox21 NR-AhR

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#450]
### Model Title

Tox21 NR-AhR (TDCommons)

### Publication

Hello @angelnagaba!

As part of your Outreachy contribution, we have assigned you the dataset "Tox21 NR-AhR" from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by angelnagaba** [comment]
Hello @GemmaTuron I have managed to import the data from the TDC however i am wondering what the **NR-AHR** Means. Kindly help me understand it as i have tried to read but i am not getting it

**Comment by Yayeks** [comment]
Hi @angelnagaba, you didn't grant access to your Colab notebook.

**Comment by angelnagaba** [comment]
hi, I have changes the permissions sorry about that

**Comment by Yayeks** [comment]
@angelnagaba, our mentor asked me this question earlier, you can take it into consideration

- Can you comment in this issue a bit more about your data? what are we modelling? how is your data distribution in classes, do you think it will be a difficult or easy task to model, and why?

**Comment by GemmaTuron** [comment]
Hi @angelnagaba !

Good job on the notebook, please take some time to explain your findings and answer the questions

**Comment by angelnagaba** [comment]
**Progress Summary**
Hello @GemmaTuron Here is a summary of my progress. 

Tox21 contains Binary classification toxicity prediction problems whereby given the drug Smiles string, we are expected to predict the toxicity of a given molecule by classifying it as either toxic or non-toxic. 
After obtaining the Dataset I used the random split to split the dataset below
1.	Train Dataset which has a size of 4584
2.	Test Dataset which has a size of 1310
3.	Validation Dataset which has a size of 655
On understanding each of them, to find out how many actives and non-active molecules were in each of the dataset. The distribution was as follows;
   **Train Dataset** had **526** active molecules and **4058** inactive molecules
   **ValidationDataset** had **59** active molecules and **596** inactive molecules
   **Train Dataset** had **183** active molecules and **1127** inactive molecules

I then used the RDKIT package to visualize the molecules as shown below; 
**Active Molecules** 
![image](https://user-images.githubusercontent.com/41667860/198578249-ec3849f4-50e8-439b-86d9-c99d98cf07ce.png)
**Inactive Molecules**
![image](https://user-images.githubusercontent.com/41667860/198578326-8149365b-574a-4b2d-b191-abc4a43d2ed0.png)

**Model Training**
I then used the Lazy-QA package to prepare my data for modelling. The Lazy QSAR library built by ersilia provides a faster way to build models. It provides a way to split the data and then fit/train the data using the training dataset. 
Model training is the primary step in machine learning, It helps us come up with a working model that can then be validated, tested and deployed. The model’s performance during training will eventually determine how well it will work when it is tested using the test dataset (dataset it has not interacted with before). Our datasets have three variables i.e. DRUGID, DRUG(Smiles) and Y. We need to use the smiles variable (input variable) and the Y variable which is our target variable. The smiles variable is a molecule and the computer is able to read only 0s and 1s (Numerical data). This implied that there was a need to convert the molecules into a readable format. However, Lazy Qsar library does this for us by automatically converting smiles to Morgan fingerprints. I achieved this by first importing the LazyQsar library from GitHub and then used its MorganBinaryClassifier to fit the model by training it. We fit the model by passing our input variable which is the Drug/Smiles Variable and our target Variable which is the Y Variable
**_model.fit(smiles_train, Y_train)_**

We then proceed and save the trained model in joblib format as _**trained_model.joblib.**_
 
**Getting Predictions for the Test dataset and Validate Dataset.** 
The trained model is then used to obtain predictions from the test and validation dataset. We split the dataset so that we can get a separate dataset to fit the model (for the model to learn the dataset) and then test the model with a dataset it has not interacted with to obtain predictions. 
The output of these predictions are probabilities and I stored the probabilities for the positive outcomes only. 
Next was to evaluate the performance of our model. After being trained and then validated and tested does it actually perform as expected?

I started by importing the necessary libraries for evaluating the model. i.e. Pandas, Scit-Learn, NumPy, sklearn, Matplotlib, seaborn. These libraries however were imported at the start to ensure I have clean code. I then used performance metric measurements like;
    •	Precision
    •	Recall
    •	F1_Score
    •	AUROC Curve and Value. 

**Threshold Value.** 
Since the outcomes of the predictions are probabilities, there is a need to convert them to integers and this required getting an optimal threshold value. Which was **0.255** in this case.

we need a way of using a single threshold to interpret the predicted probabilities. This can be achieved by mapping all values equal to or greater than the threshold to 1 and all values less than the threshold to 0. This was achieved by defining a function that takes the probabilities/predictions and threshold as an argument and return an array of integers in {0, 1}. We proceed to evaluate each threshold value using the F1_score and then locate the array index with the largest score and use it as our optimal threshold value. I used this optimal threshold value to convert the predictions to discrete values making it easy to test the performance of the model. 

**Model Performance;**
**1.	Confusion Matrix**
![image](https://user-images.githubusercontent.com/41667860/198578660-73c7abf4-aa62-4bfc-8d79-d45ca237dedd.png)
          i.	548 Molecules being classified by the model as being Non-toxic are actually Toxic (True Negatives)
          ii.	37 molecules are being classified as toxic and are actually Toxic (True Positives)
          iii.	22 molecules are being classified as Non-toxic yet they are actually Toxic (False Negative)
          iv.	48 molecules were classified as Toxic yet they are actually non-toxic. (False Positive)
          
This could be attributed to the imbalanced datasets that were driven due to use of random split method. This could be solved by using the Scaffold Split which mainly bases on the structure of the molecules. 
 

**2.	AUROC Curve and Value Performance.** 
 
![image](https://user-images.githubusercontent.com/41667860/198579035-d1fdd291-7286-42a4-a03f-54473869c3d4.png)

The ROC Curve is a useful diagnostic tool that helps us visualize how well our morganbinaryclassifier is performing. 

AUROC SCORE OF 0.89, Implies that this is a rich model and likely to predict positives as positives and negatives as negatives based on The judging criteria The Higher the AUC, the better the model.

**The Validation Dataset model performance**
1.	Precision for Validation set. 
Precision is the positive predictive value. This helps us evaluate: When the model predicted positives was it actually positive
**Precision score = 0.435294**
2.	Recall 
Out of all the possible positives, how many did the model correctly identify
**Recall score: 0.627119**
3.	F1_score 
**F1 Score: 0.513889** 

**Test Dataset Model performance**
    1.Precision: 0.513889
    2.Recall: 0.595628
    3.f1_score: 0.558974

Here is My [colabNotebook ](https://colab.research.google.com/drive/1nSAcj_4o5HRU9YmA-BEzlGOrDaDVM_gW?usp=sharing)for your reference
****N;B** I am still reading more about how to interprete these scores and improve them.**

**Comment by angelnagaba** [comment]
1. The dataset is abit imbalanced with more inactives than active molecules. this could affect our model being trained as it will be exposed to more inactive molecules making it hard for it to recognize active molecules which in turn affects the accuracy/performance of the model. 
2. We are modeling a Binary classification problem where outcomes are binary(0 or 1). 
3. Data Distribution: In general the dataset has an imbalanced number of active and inactive molecules. In general, the dataset has more inactive molecules than active molecules. this will make it hard to accurately train the model as this would require careful splitting of the data.

**Comment by angelnagaba** [comment]
> @angelnagaba, our mentor asked me this question earlier, you can take it into consideration
> 
> * Can you comment in this issue a bit more about your data? what are we modelling? how is your data distribution in classes, do you think it will be a difficult or easy task to model, and why?

1. The dataset is abit imbalanced with more inactives than active molecules. this could affect our model being trained as it will be exposed to more inactive molecules making it hard for it to recognize active molecules which in turn affects the accuracy/performance of the model. 
2. We are modeling a Binary classification problem where outcomes are binary(0 or 1). 
3. Data Distribution: In general the dataset has an imbalanced number of active and inactive molecules. In general, the dataset has more inactive molecules than active molecules. this will make it hard to accurately train the model as this would require careful splitting of the data.

**Comment by GemmaTuron** [comment]
Hi @angelnagaba !

Good job, my comments would be to try and increase training time to see if we can get a better precision and recall. 
I'll close this as completed and you can go onto listing it in your outreachy contributions and start preparing the final application, thanks!
